Add xl!unsetprefix command, better support for wine, and Retainer Sale support

reiichi001 opened the pull request:
Users weren't sure how to unset a channel prefix as it wasn't documented anywhere. Rather than pitch the "none" feature I found in the code, I added a dedicated command and help for this as I think that will be easier on users to figure out.

For some reason, the linq code doesn't seem to like working when something's been removed, but there's still something left. I don't know why. If this is something I can fix pre-commit, lemme know in comments and I'll re-test!

reiichi001 committed:
Add xl!unsetprefix command

Users weren't sure how to unset a channel prefix as it wasn't documented anywhere. Rather than pitch the "none" feature I found in the code, I added a dedicated command and help for this as I think that will be easier on users to figure out.

For some reason, the linq code doesn't seem to like working when something's been removed, but there's still something left. I don't know why.

reiichi001 committed:
Make the plugin wine friendly

According to https://github.com/goatcorp/Dalamud/issues/135 this would make the discord bot plugin work for Wine/Linux users.

reiichi001 committed:
Add support for retainer item sales

Very basic. Not as fancy as the old plugin yet, but a nice first attempt.

Copied some code over from the old bot and made some adjustments. Currently just relaying the full message with webhook using the item name/icon.

reiichi001 committed:
xl!help should respect bot prefix

Swap out hardcoded `!` for the bot's configured prefix in all help commands and examples.

reiichi001 committed:
Get the Bridge actually running on start

Use Task.Run() to force DiscordHandler to start because it wouldn't otherwise.

Also avoid calling XIVAPI if sendername/senderworld are empty to cut down on errors.

reiichi001 committed:
Fix outgoing tells and silence some types

Outgoing tells would make XIVAPI mad because they would always resolve to the recipient's world, not the sender's. Also added some handling to ignore certain types that will never have a player link so the log is more readable and doesn't overflow with shit.

reiichi001 committed:
Add missing permissions

Default join lacked permission to send an embed. 

Also added permissions for external emoji and reactions, as a preventative measure since bots use those too, and we might want to in the future.

reiichi001 committed:
Silence more unneded chat types and fix list

Discord was interpretting the first line of the aggregated chat types as a syntax highlighting type, which would skip the first entry. Sometimes. Added a newline before it so avoid this as a good practice.

I've also adjusted some more chattypes to skip checking for player link as they can't/won't have one afaik. Keeps the log tidy.

reiichi001 committed:
Silence bNPCs too

reiichi001 committed:
Preliminary support for duplicate message ignore

Also removed an old debug log to keep the dalamud.txt log tidy.

reiichi001 committed:
Make build.yml

Copied from https://github.com/BiscuitUmU/FantasyPlayer

reiichi001 committed:
Make release.yml

Copied from https://github.com/BiscuitUmU/FantasyPlayer

reiichi001 committed:
Fix specialcharhandler typo

Pointed out by Jajali Jali

reiichi001 committed:
Update build.yml

reiichi001 committed:
Update build.yml

reiichi001 committed:
Merge remote-tracking branch 'origin/master'

goaaats commented on Dalamud.DiscordBridge/DiscordHandler.cs:
```diff
@@ -418,6 +482,49 @@ private async Task<bool> EnsureOwner(IUser user, ISocketMessageChannel errorMess
+            PluginLog.Information($"Retainer sold item: {itemId}");
+            try
+            {
+                ItemResult res = XivApiClient.GetItem(itemId).GetAwaiter().GetResult();
```
Don't use XIVAPI for this, especially not in DiscordHandlers - I'd grab it from Lumina in DiscordMessageQueue

reiichi001 replied:
Is there a better way to grab the icon? That's all I really needed xivapi for. This was largely copied from the old bot and I could adjust the function to have the MessageQueue send over the item name, cutting things down to really just a need for an icon. 

Based on some quick testing, the icon ushort that Lumina provides from an item doesn't match xivapi URLs, so I'm not sure if we can just hardcode one programmatically or if I'm just looking at the wrong thing.

goaaats commented on Dalamud.DiscordBridge/Model/QueuedRetainerItemSaleEvent.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class QueuedRetainerItemSaleEvent : QueuedXivEvent
+    {
+        public SeString Message { get; set; }
```
add the Item row here, populate it in DiscordMessageQueue and pass it to DiscordHandlers

reiichi001 replied:
DiscordMessageQueue is already handling it https://github.com/reiichi001/Dalamud.DiscordBridge/blob/master/Dalamud.DiscordBridge/DiscordMessageQueue.cs#L88 

In DiscordMessageQueue, I currently process the Item payload because there are multiple things we'd want to get out of it anyways, like number of items (to do) and whether it was HQ (sent, but possibly not needed as the message has this already handled), which is then sent to the handler.

goaaats commented on Dalamud.DiscordBridge/Plugin.cs:
```diff
@@ -27,8 +29,26 @@ public void Initialize(DalamudPluginInterface pluginInterface)
+            });
+
+            // try to force-reload it
+            if (!this.Discord.IsConnected)
```
Does this actually do anything? Try to breakpoint - you'll likely see that this gets ran every time, since the Task.Run from before didn't even get to run yet at that point.

The first one may be enough.

reiichi001 replied:
It can be cut down a bit. Some of that is leftovers from debugging.

With just `Task t = this.Discord.Start();`, the bot never actually starts when Dalamud loads the plugin normally. (See multiple users reporting this in issues on discord). Using Task.Run() solved it, although the check to try again can be safely removed.

reiichi001 committed:
Remove some cruft

Moved xivapi call to get item icon over to DiscordMessagequeue and removed some leftover debug code that was doubling up on things we didn't need.

Removed some parameters for item send messages we no longer need because they're handled over in the message queue anyways.

reiichi001 committed:
Generate link for xivapi icon instead of fetching it

Reduces XIVAPI calls because we can determine the iconURL ourselves instead of getting the item and then getting the link.

reiichi001 committed:
Make duplicate message log easy to get

reiichi001 committed:
Fix missing senderWorld for cross world standardemote

Have I mentioned I hate playerlink handling? Because I do. This parses out a senderWorld for cross-world standardemotes because those don't get a playerlink. Trims off the player name since we still get that from the sender. Then splits on whitespace to get a world name. 

Also fixes some stupidity with chat duplication where it always ran on 0ms when that should disable it instead.

reiichi001 committed:
Handle special emotestandard case

Make parsing for emotestandard handle cases where it appends " 's " onto the character name.

reiichi001 committed:
Use message's playerlink instead of parsing

Thanks Haplo for nicer solution.

Also set some different player=null things for some system/game messages we don't plan to log.

reiichi001 committed:
Add support for npctalk chat type.

Add in support for npctalk chat type (61).

Also revise the xivapi character search to ignore echo and npc messages to cut down on API calls.

reiichi001 committed:
Add npcannounce type and channel

reiichi001 committed:
Version bump - compilation hints

reiichi001 committed:
Update build.yml

reiichi001 committed:
Update release.yml

reiichi001 committed:
Update build.yml

reiichi001 committed:
Update release.yml

reiichi001 committed:
Fix hints

reiichi001 committed:
Merge remote-tracking branch 'origin/master'

reiichi001 committed:
Silence some unwanted chat types

Also bumps to version 1.0.1.1 for updates.

reiichi001 committed:
Update for API 3

Changes for Dalamud API3.

Also removing WS4Net because it was causing problems. I never heard from any Wine users if they were impacted, so dependencies could probably be cleaned up a bit if there's any cruft.

reiichi001 committed:
Fix build actions

This will still be a broken compile because it's for staging. I'll do another release after patch 5.5 comes out to update the repo (or actually deploy to DalamudPlugins/testing)

reiichi001 committed:
Switch to DalamudPackager and update links

reiichi001 committed:
Add SystemMessage and GM Chat types

reiichi001 commented:
Closing PR since I'm defacto maintainer for now
